File: python/excercise236/lowest_common_ancestor_of_a_binary_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x, left=None, right=None):
        self.val = x
        self.left = left
        self.right = right


def lowest_common_ancestor(root: 'TreeNode', p: 'TreeNode', q: 'TreeNode'):
    if root is None or root.val == p.val or root.val == q.val:
        return root
    left = lowest_common_ancestor(root.left, p, q)
    right = lowest_common_ancestor(root.right, p, q)

    if left and right:
        return root

    if left:
        return left

    return right

File: go/excercise236/lowest_common_ancestor_of_a_binary_tree.go
package excercise236

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if root == nil || root == p || root == q {
		return root
	}

	left := lowestCommonAncestor(root.Left, p, q)
	right := lowestCommonAncestor(root.Right, p, q)

	if left == nil {
		// 左子树为空，肯定都在右子树上
		return right
	}

	if right == nil {
		// 右子树为空，肯定都在左子树上
		return left
	}

	// 左右子树都不为空，一个在左子树一个在右子树，所以root就是他们的最近公共祖先节点。
	return root
}
